cmd/ocp-build-data-enforcer/main.go

```diff
@@ -0,0 +1,304 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/openshift/imagebuilder"
+	dockercmd "github.com/openshift/imagebuilder/dockerfile/command"
+	"github.com/pmezard/go-difflib/difflib"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/github"
+)
+
+type options struct {
+	ocpBuildDataRepoDir string
+	majorMinor          majorMinor
+}
+
+func gatherOptions() *options {
+	o := &options{}
+	flag.StringVar(&o.ocpBuildDataRepoDir, "ocp-build-data-repo-dir", "../ocp-build-data", "The directory in which the ocp-build-data reposity is")
+	flag.StringVar(&o.majorMinor.minor, "minor", "6", "The minor version to target")
+	flag.Parse()
+	return o
+}
+func main() {
+	opts := gatherOptions()
+	opts.majorMinor.major = "4"
+
+	configsUnverified, err := gatherAllOCPImageConfigs(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to gather all ocp image configs")
+	}
+
+	streamMap, err := readStreamMap(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to read streamMap")
+	}
+
+	groupYAML, err := readGroupYAML(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Faild to read groupYAML")
+	}
+
+	var errs []error
+	var configs []ocpImageConfig
+	for _, cfg := range configsUnverified {
+		if err := cfg.validate(); err != nil {
+			errs = append(errs, fmt.Errorf("error validating %s: %w", cfg.SourceFileName, err))
+			continue
+		}
+		if err := dereferenceConfig(&cfg, streamMap, groupYAML); err != nil {
+			errs = append(errs, fmt.Errorf("failed dereferencing config for %s: %w", cfg.SourceFileName, err))
+			continue
+		}
+		configs = append(configs, cfg)
+	}
+
+	errGroup := &errgroup.Group{}
+	for idx := range configs {
+		idx := idx
```

Comment:
This is super opaque -- why not 

```go
for i := range configs {
    errGroup.Go(func(config ocpImageConfig) error {
        return processDockerfile(config)
    }(configs[i]))
}
```

Reply:
because errGroup.Go takes a `func() error` as first argument: https://godoc.org/golang.org/x/sync/errgroup#Group.Go

Reply:
ew. Maybe a double closure? Or some comment -- the code as written is non-obvious

Reply:
Sorry, what is non-obvious about this?

Reply:
I find the whole passing it in as argument to copy much harder to read than this